src/poll.c

```diff
@@ -0,0 +1,356 @@
+struct poll_data {
+	char *path;
+	uint32_t path_hash;
+	// Last update timestamp
+	struct timeval *tv;
+	int num_checker;
+} *poll_data_array;
+
+int poll_num = 0; // Size of poll_fd_array|poll_data_array
+
+// Every registrator holds a void pointer to one of these.
+struct poll_check {
+	int poll_array_index;
+	// Last check timestamp.
+	struct timeval tv;
+};
+
+pthread_t poll_thread;
+
+// ---------------------------------------------------------------------------
+// poll locks
+
+// Lock poll data structure (poll_fd_array, poll_data_array and poll_num)
+pthread_mutex_t poll_array_mtx = PTHREAD_MUTEX_INITIALIZER;
+
+static inline int poll_array_lock() {
+	if(pthread_mutex_lock(&poll_array_mtx) == 0) {
+		return 0;
+	} else {
+		error("Unable to lock poll file list");
+		return -1;
+	}
+}
+
+static inline int poll_array_unlock() {
+	if(pthread_mutex_unlock(&poll_array_mtx) == 0) {
+		return 0;
+	} else {
+		error("Unable to unlock poll file list");
+		return -1;
+	}
+}
+
+// ---------------------------------------------------------------------------
+// internal signal handler used to interrupt poll.
+void poll_handler(int signo)
+{
+	if(signo){} // Remove compiler warning
+}
+
+// ---------------------------------------------------------------------------
+// internal methods
+int poll_time_update_nolock(struct timeval *tv) {
+	int retv = gettimeofday(tv, NULL);
+	if(retv != 0) {
+		switch(errno) {
+			case EFAULT:
+			case EINVAL:
+				error("Could not get current time");
+				break;
+			case EPERM:
+				error("CAP_SYS_TIME capability is required to get current time");
+				break;
+		}
+	}
+	return retv;
+}
+
+// if bigger is less than lower we return 0.
+unsigned long long poll_time_difference_nolock(struct timeval *bigger, struct timeval *lower) {
+	struct timeval diff;
+	if(timercmp(bigger, lower, >)) {
+		timersub(bigger, lower, &diff);
+		return (diff.tv_sec * 1000000) + diff.tv_usec;
+	} else {
+		return 0;
+	}
+}
+
+int poll_array_search_nolock(char *path, int type) {
+	uint32_t hash = simple_hash(path);
+	int i;
+	for(i = 0; i < poll_num; i++) {
+		if(poll_fd_array[i].fd < 0 && 
+				poll_data_array[i].path_hash == hash &&
+				poll_fd_array[i].events == type &&
+				strcmp(poll_data_array[i].path, path) == 0)
+			return i;
+	}
+	return -1;
+}
+
+int poll_array_find_first_deprecated_nolock() {
+	int index = 0;
+	for(index; index<poll_num; index++)
+		if(poll_fd_array[index].fd < 0)
+			return index;
+	return -1;
+}
+
+void poll_array_remove_nolock(int index) {
+	if(index >= poll_num) return;
+
+	if(poll_fd_array[index].fd >= 0)
+		if(close(poll_fd_array[index].fd) != 0)
+			error("Failed to proper close file descriptor %d", poll_fd_array[index].fd);
+	poll_fd_array[index].fd = -1;	
+}
+
+int poll_array_add_nolock(char *path, int type) {
+	int index;
+
+	index = poll_array_find_first_deprecated_nolock();
+	if(index >= 0) {
+		poll_num++;
+		poll_fd_array = realloc(poll_fd_array, sizeof(struct pollfd) * poll_num);
```

Comment:
also memory allocation functions now use the same ending with `z`. This takes care of failures, so you don't need to check for NULL responses.

i.e. `mallocz()`, `callocz()`, `reallocz()`, `strdupz()`, `freez()`.